data_structures/linked-list/linked_list.py

```diff
@@ -39,23 +39,31 @@ def find(self, val):
             search = search._next
         return False
 
+    # I think these should work, but I'm trying to figure out how to write tests for them
+
     def append(self, val):
         """Append a node to the end of the list."""
         search = self.head
         while search._next is not None:
             search = search._next
         search._next = Node(val)
 
-    # def insert_before(self, val, new_val):
-    #     """Insert a new node before the matching value."""
-    #     search = self.head
-    #     while search._next.val != val:
-    #         search = search._next
-    #     search = Node(new_val, search._next)
-
-    # def insert_after(self, val, new_val):
-    #     """Insert a new node after the matching value."""
-    #     search = self.head
-    #     while search.val != val:
-    #         search = search._next
-    #     search.next = Node(new_val, search._next)
+    def insert_before(self, val, new_val):
+        """Insert a new node before the matching value."""
+        try:
+            search = self.head
+            while search._next.val != val:
+                search = search._next
+            search = Node(new_val, search._next)
+        except (ValueError, KeyError):
+            print('That is not a valid value!')
```

Comment:
You should not be just printing that the value is invalid. The spec calls for an Exception, thus you should be raising an exception yourself. Something along the lines of
```python
raise ValueError('That is not a valid value!')
```

Reply:
The nice thing about this is that now you can test to check that your method is raising the appropriate exception when the value is not in the list.